lib/charms/grafana_k8s/v0/grafana_source.py

```
@@ -551,6 +563,13 @@ def _on_grafana_peer_changed(self, _: RelationChangedEvent) -> None:
         self.on.sources_changed.emit()  # pyright: ignore
         self.on.sources_to_delete_changed.emit()  # pyright: ignore
 
+    def _publish_source_uids(self, rel: Relation, uids: Dict[str, str]):
+        """Share the datasource UIDs back to the datasources.
+
+        Assumes only leader unit will call this method
+        """
+        rel.data[self._charm.app]["datasource_uids"] = json.dumps(uids)
```

Should we have one more leader check just in case someone misuses the method?

---

hmm I think in this case, I'd want it error out loudly rather than silently not doing anything

---

plus its intended for private use